splink/cluster_metrics.py

```diff
@@ -0,0 +1,65 @@
+from splink.input_column import InputColumn
+
+
+def _size_density_sql(
+    df_predict, df_clustered, threshold_match_probability, _unique_id_col
+):
+    """Generates sql for computing cluster size and density at a given threshold.
+
+    Args:
+        df_predict (SplinkDataFrame): The results of `linker.predict()`
+        df_clustered (SplinkDataFrame): The outputs of
+                `linker.cluster_pairwise_predictions_at_threshold()`
+        threshold_match_probability (float): Filter the pairwise match
+            predictions to include only pairwise comparisons with a
+            match_probability above this threshold.
+        _unique_id_col (string): name of unique id column in settings dict
+
+    Returns:
+        sql string for computing cluster size and density
+    """
+
+    # Get physical table names from Splink dataframes
+    edges_table = df_predict.physical_name
+    clusters_table = df_clustered.physical_name
+
+    input_col = InputColumn(_unique_id_col)
+    unique_id_col_l = input_col.name_l()
+
+    sqls = []
+    sql = f"""
+        SELECT
+            {unique_id_col_l},
+            COUNT(*) AS count_edges
+        FROM {edges_table}
+        WHERE match_probability >= {threshold_match_probability}
+        GROUP BY {unique_id_col_l}
+    """
+
+    sql = {"sql": sql, "output_table_name": "__splink__count_edges"}
+    sqls.append(sql)
+
+    sql = f"""
+        SELECT
+            c.cluster_id,
+            count(*) AS n_nodes,
+            sum(e.count_edges) AS n_edges
+        FROM {clusters_table} AS c
+        LEFT JOIN __splink__count_edges e ON c.{_unique_id_col} = e.{unique_id_col_l}
```

Comment:
I think this join needs to account for the source dataset


<details>
<summary>
Reprex
</summary>

```python
import pandas as pd
from IPython.display import display

from splink.duckdb.duckdb_comparison_library import (
    exact_match,
)
from splink.duckdb.duckdb_linker import DuckDBLinker

settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "link_only",
    "comparisons": [
        exact_match("first_name"),
        exact_match("surname"),
        exact_match("dob"),
    ],
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
}


df_1 = [
    {"unique_id": 1, "first_name": "Tom", "surname": "Fox", "dob": "1980-01-01"},
    {"unique_id": 2, "first_name": "Amy", "surname": "Lee", "dob": "1980-01-01"},
]


df_2 = [
    {"unique_id": 1, "first_name": "Bob", "surname": "Ray", "dob": "1999-09-22"},
    {"unique_id": 2, "first_name": "Amy", "surname": "Lee", "dob": "1980-01-01"},
]

df_1 = pd.DataFrame(df_1)
df_2 = pd.DataFrame(df_2)

linker = DuckDBLinker(
    [df_1, df_2], settings, input_table_aliases=["df_left", "df_right"]
)


df_predict = linker.predict()
display(df_predict.as_pandas_dataframe())


df_clustered = linker.cluster_pairwise_predictions_at_threshold(df_predict, 0.9)
display(df_clustered.as_pandas_dataframe().sort_values("cluster_id"))

linker.debug_mode=True
linker._compute_cluster_metrics(df_predict, df_clustered, 0.9).as_pandas_dataframe()

```

</details>

Reply:
![Uploading image.png…]()

Reply:
There are a few Splink functions that help with this.

Maybe [_unique_id_input_columns](https://github.com/moj-analytical-services/splink/blob/42de8da5e9ce50f26c6c7de287d167c421e08dd5/splink/settings.py#L171)

or possibly here
https://github.com/moj-analytical-services/splink/blob/a9f5424995dc3e4473e8df0c7f97408454c178c0/splink/unique_id_concat.py#L1